### PR TITLE
Google spark operator

### PR DESCRIPTION
**Pull request by guimou**
First implementation.
Installation in ODH namespace, listens to all namespaces.

### CODE REVIEW

File: google-spark-operator/operator/base/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        prometheus.io/path: /metrics
+        prometheus.io/port: "10254"
+        prometheus.io/scrape: "true"
+      labels:
+        app.kubernetes.io/instance: spark-operator
+        app.kubernetes.io/name: spark-operator
+    spec:
+      containers:
+        - name: sparkoperator
+          image: gcr.io/spark-operator/spark-operator:v1beta2-1.2.3-3.1.1
```
Comment (rimolive) [inline google-spark-operator/operator/base/deployment.yaml:25]: Can we move to our internal quay.io repo?

File: google-spark-operator/operator/base/webhook-svc.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+kind: Service
+apiVersion: v1
+metadata:
+  name: spark-operator-webhook
+  namespace: spark-operator
```
Comment (rimolive) [inline google-spark-operator/operator/base/webhook-svc.yaml:5]: Remove this line

File: google-spark-operator/operator/base/webhook-svc.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  labels:
+    app.kubernetes.io/instance: spark-operator
+    app.kubernetes.io/name: spark-operator
+spec:
+  ports:
+    - name: webhook
+      protocol: TCP
+      port: 443
```
Comment (rimolive) [inline google-spark-operator/operator/base/webhook-svc.yaml:13]: Why port is 443 and redirects to 8080? Can we at least use a secure port inside the cluster too?

File: google-spark-operator/operator/base/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+commonLabels:
+  app.kubernetes.io/name: sparkoperator
+  kustomize.component: spark-operator
+images:
+- name: gcr.io/spark-operator/spark-operator
```
Comment (rimolive) [inline google-spark-operator/operator/base/kustomization.yaml:6]: Can we move to quay.io?

File: google-spark-operator/operator/base/Kube-descriptor.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+appVersion: v1beta2-1.2.3-3.1.1
+description: Spark operator based on https://github.com/GoogleCloudPlatform/spark-on-k8s-operator
+home: https://github.com/opendatahub-io/odh-manifests/google-spark-operator
+keywords:
+- spark
+kubeVersion: '>=1.8.0-0'
+name: sparkoperator
+version: 0.0.1
```
Comment (rimolive) [inline google-spark-operator/operator/base/Kube-descriptor.yaml:9]: Is this file required? At least it is not listed in kustomize.yaml

**Comment by HumairAK** [comment]
When deploying this in our operate-first cluster I'm seeing the following errors: 

from ` spark-operator-webhook-init-n9zzc` pod: 
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "secrets is forbidden: User \"system:serviceaccount:opf-google-spark-operator:spark-operator\" cannot create resource \"secrets\" in API group \"\" in the namespace \"opf-google-spark-operator\"",
  "reason": "Forbidden",
  "details": {
    "kind": "secrets"
  },
  "code": 403
}
Failed creating secret.
```
The above is resulting in a crashloopbackoff.

There's also this pod: `spark-operator-867df57854-lwkdt` which is failing to create with the following event error: 

`MountVolume.SetUp failed for volume "webhook-certs" : secret "spark-operator-webhook-certs" not found`

the kfdef spec: 

```yaml
spec:
  applications:
    - kustomizeConfig:
        repoRef:
          name: guimou
          path: google-spark-operator/operator
      name: google-spark-operator
  repos:
    - name: manifests
      uri: 'https://github.com/opendatahub-io/odh-manifests/tarball/v1.0.11'
    - name: guimou
      uri: 'https://github.com/guimou/odh-manifests/tarball/spark-operator'
  version: v1.0.11

```

**Comment by guimou** [comment]
Hold on before testing further. I'm implementing some of the changes asked by @rimolive, and it's still wip. I'll update here when it's ready.

File: google-spark-operator/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+# Google Spark operator
+
+[Kubernetes Operator for Spark](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator) contains one component, the [spark operator](#spark-operator), that deploys [Apache Spark](https://spark.apache.org) clusters for large-scale data processing.
+
+By default, the operator will be deployed in the namespace where Open Data Hub is installed, but will listen for SparkApplication and Scheduled SparkApplication objects in all namespaces.
+
+If workloads are deployed outside of the Open Data Hub namespace, it may be necessary to create a dedicated ServiceAccount to be uses with those workloads, along with necessary Role and Rolebindings. You can refer to the following files for examples:
```
[the remark this answers is not in the document]
Comment (guimou) [reply google-spark-operator/README.md]: Corrected

File: google-spark-operator/operator/base/deployment.yaml (hunk 1 above)
Comment (guimou) [reply google-spark-operator/operator/base/deployment.yaml:29]: I don't have access to the quay repo to do it. But as soon as the image is copied there I can make the changes.

File: google-spark-operator/operator/base/spark-sa.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: spark
+  namespace: opendatahub
```
[the remark this answers is not in the document]
Comment (guimou) [reply google-spark-operator/operator/base/spark-sa.yaml]: Done

File: google-spark-operator/operator/base/webhook-svc.yaml (hunk 3 above)
Comment (guimou) [reply google-spark-operator/operator/base/webhook-svc.yaml:13]: It's funny because in fact it's a TLS connection that is opened in the code, even if it's on port 8080. Therefore the certificates that are needed to connect to the Service.
But to avoid confusion I changed it for 8443.

File: google-spark-operator/operator/base/webhook-certs-job.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        job-name: spark-operator-webhook-init
+    spec:
+      containers:
+        - name: main
+          env:
+            - name: NAMESPACE
+              valueFrom: 
+                fieldRef:
+                  fieldPath: metadata.namespace
+          image: 'gcr.io/spark-operator/spark-operator:v1beta2-1.2.3-3.1.1'
```
[the remark this answers is not in the document]
Comment (guimou) [reply google-spark-operator/operator/base/webhook-certs-job.yaml:21]: Same as previous, as soon as image is available I'll make the changes.

File: google-spark-operator/operator/base/webhook-svc.yaml (hunk 2 above)
Comment (guimou) [reply google-spark-operator/operator/base/webhook-svc.yaml]: Done

File: google-spark-operator/operator/base/kustomization.yaml (hunk 4 above)
Comment (guimou) [reply google-spark-operator/operator/base/kustomization.yaml:6]: Same...

File: google-spark-operator/operator/base/Kube-descriptor.yaml (hunk 5 above)
Comment (guimou) [reply google-spark-operator/operator/base/Kube-descriptor.yaml]: File removed

**Comment by guimou** [comment]
> When deploying this in our operate-first cluster I'm seeing the following errors:
> 
> from ` spark-operator-webhook-init-n9zzc` pod:
> ....
@HumairAK 
All those secrets mismatches have been corrected.

File: google-spark-operator/operator/base/deployment.yaml (hunk 1 above)
Comment (crobby) [reply google-spark-operator/operator/base/deployment.yaml:29]: @guimou I've added you to the opendatahub quay org.  you should be able to push away.

Comment (guimou) [reply google-spark-operator/operator/base/deployment.yaml:29]: Thanks @crobby! Is that OK if I name it google-spark-operator to differentiate if from radanalytics? Other suggested name?

Comment (crobby) [reply google-spark-operator/operator/base/deployment.yaml:29]: That name sounds fine to me.

Comment (guimou) [reply google-spark-operator/operator/base/deployment.yaml]: Changed for ODH quay repo

File: google-spark-operator/operator/base/webhook-certs-job.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (guimou) [reply google-spark-operator/operator/base/webhook-certs-job.yaml]: Done

**Comment by guimou** [comment]
All changes done and retested.

File: google-spark-operator/operator/base/params.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+varReference:
+- path: subjects/namespace
+  kind: ClusterRoleBinding
+- path: subjects/namespace
+  kind: RoleBinding
+- path: spec/template/spec/containers/args
+  kind: Deployment
```
Comment (crobby) [inline google-spark-operator/operator/base/params.yaml:7]: small nit:  newline

Comment (guimou) [reply google-spark-operator/operator/base/params.yaml]: There you are! :-)

**Comment by 4n4nd** [comment]
@guimou I deployed your branch in operate first and tried to run an example spark application ([this one](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/examples/spark-pi-configmap.yaml)). It created a pod but it failed.
The CR status was:
```yaml
status:
  applicationState:
    errorMessage: 'driver container failed with ExitCode: 1, Reason: Error'
    state: FAILED
  submissionID: 437f2a2b-935a-4348-aea8-4b0788d1f11f
  sparkApplicationId: spark-8f4f4c79201844f8a5721b7e2aaa59bf
  terminationTime: '2021-05-27T20:53:04Z'
  submissionAttempts: 1
  driverInfo:
    podName: spark-pi-driver
    webUIAddress: '172.30.55.75:4040'
    webUIPort: 4040
    webUIServiceName: spark-pi-ui-svc
  lastSubmissionAttemptTime: '2021-05-27T20:52:24Z'
  executionAttempts: 1
  executorState:
    spark-pi-293fa079af9a12db-exec-1: FAILED
```

And in the driver pod logs, I found this error log
```
Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://kubernetes.default.svc/api/v1/namespaces/opf-google-spark-operator/configmaps. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. configmaps is forbidden: User "system:serviceaccount:opf-google-spark-operator:spark" cannot create resource "configmaps" in API group "" in the namespace "opf-google-spark-operator".
```

Also, the Spark App was only submitted when the App was created in the same namespace as the operator. It did not work in a different namespace.

**Comment by tloubrieu-jpl** [comment]
Hello @vpavlin @anishasthana, is this pull request pending because of the missing reviews ? Sorry to poke at you, I am asking since we need this pull request for our project. Thanks

**Comment by guimou** [comment]
> @guimou I deployed your branch in operate first and tried to run an example spark application ([this one](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/examples/spark-pi-configmap.yaml)). It created a pod but it failed.
> And in the driver pod logs, I found this error log
> 
> ```
> Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://kubernetes.default.svc/api/v1/namespaces/opf-google-spark-operator/configmaps. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. configmaps is forbidden: User "system:serviceaccount:opf-google-spark-operator:spark" cannot create resource "configmaps" in API group "" in the namespace "opf-google-spark-operator".
> ```
@4n4nd This example is weird as it tries to create/use a ConfiMap from scratch. Therefore the operator would need the right to do that (the other examples are also definitely not meant for OpenShift. The basic spark-pi one uses a hostPath, which is a no-no for us). So three solutions:
- create the configmap in advance (athough I'm not sure it it needs write access, it which case it may fail).
- change the volume for another suitable type: PV, EmptyDir,...
- add the ability for the Spark role to do create CMs. In which case we would diverge from the standard operator behaviour as this is not part of the original role.

> Also, the Spark App was only submitted when the App was created in the same namespace as the operator. It did not work in a different namespace.
Strange, not supposed to. In this other namespace, did you create the ServiceAccount with the right role attached to it? Like the Spark SA that is created when deploying? Because the operator is watching all the namespaces, but it needs a ServiceAccount there to run the workload. As we don't know the namespace(s) the user would want to use apart from where it's deployed, we cannot do this creation automatically...
If everything is setup like this, do you have any log from the operator pod?

**Comment by 4n4nd** [comment]
> This example is weird as it tries to create/use a ConfiMap from scratch. Therefore the operator would need the right to do that (the other examples are also definitely not meant for OpenShift.

Got it, I just wanted to be able to test the operator so I tried to find an example app in the upstream repo.



> did you create the ServiceAccount with the right role attached to it? Like the Spark SA that is created when deploying? Because the operator is watching all the namespaces, but it needs a ServiceAccount there to run the workload.

ah that would make sense, I never created a separate `spark` SA in the namespace, I'll give that a shot :+1: 

> do you have any log from the operator pod?
I do not, but I will try it again and try to get the pod logs.

Thanks!

**Comment by anishasthana** [comment]
@guimou Something looks _super_ wrong with the commits :sweat_smile:

**Comment by guimou** [comment]
Closing to follow new component procedure, will resubmit at some point.
